### ISSUE TITLE
Setting SigningKey to the SignedXml Object

### ISSUE THREAD

**Issue by saleelpk95** [jmhardison/Example-AzureKeyVaultHSM-XMLSigning#40]
This is more of a doubt than issue. Computing the signature using Key Vault makes sense only if the private key need not be fetched explicitly. I see in the code that the private key is being set but with the additional comment that says "the actual private key is never returned by Azure KeyVault HSM".
Is there a way this can be done without setting the private key at all?

**Comment by jmhardison** [comment]
The private key isn't set, but a bundle is retrieved that includes the public details of the private key. If you dig into the documentation for that particular method, it will provide the details for you. The details of that are required in order to sign the document, and ultimately verify that signing with the public details that are embedded. 

And to your point, yes this only makes sense in an environment where it is required that an HSM is used. This was simply build to show that process with Azure KVHSM. Most other HSM's provided an SDK that already supplied the required methods that would perform the hash and validation against their HSM appliances. KeyVault did not have that when I put this together originally. Not sure if they do now, as I haven't checked.

So again, the code isn't setting a private key, it's setting the public details of the private key, which the public RSA is embedded into the document. `signedXml.SigningKey = key;`
 The custom signer sends the resulting hash of the file after embedding the details, to Azure KeyVault to compute the signature (where the private key is) and embed the results.

With all that data, the file can then be validated using the standard crypto validators in the XML Document methods.

**Comment by jmhardison** [comment]
I should also be clear, this only applies to the HSM deployed KeyVault. An HSM (azure or otherwise) generates the private key in itself in hardware. The private key never leaves that hardware, and the software must make calls to that hardware to have it compute signatures.

Loosely:

Consumer App -> Azure KV HSM API -> HSM Hardware | hardware does computation with private key and returns response back.

**Comment by saleelpk95** [comment]
Makes sense. 
On another note, I realized a bit late that I can't use HSM deployed key vault. However for the signing process, I would prefer not to fetch the private key from the KeyVault to build the SignedXml. 
Does the certificate bundle concept work with non-HSM KeyVault too? As of what I understand from your comments, it doesn't. Just wanted to clarify the same.

**Comment by jmhardison** [comment]
Ah ok I get where you are going with it. Sorry I was trying to be clear bout the intent of this example. Functionally, an Azure KV and Azure KV HSM should 'work' the same for you its just a distinction of where the key processing happens. Either it's in software (non hsm) or its in hardware (hsm).

If we got back to the original loose depiction, in a non-hsm backed keyvault it would be:
Consumer App -> Azure KV API -> SOFTWARE Processing | Software does computation with private key and returns response back.

(The at rest key that Microsoft uses to protect your instance is stored in HSM, but the keys you create and use are not stored or processed in hardware.)

Jeopardy knowledge aside. Azure KeyVault does not support key export operations, so there isn't a method you can call (including the method in this example) that would return the private portions of the key. It does allow for backup in the UI portal, but is encrypted with an Azure KV instance key you don't have, so would not be able to decrypt it.

While method naming here makes it feel like you are getting the private portion of a key, you are not. Only the public is returned in either case. When you submit the call for signing to KV, your processing will happen in either Hardware or Software depending on if it is or is not an HSM.

TLDR; You are good. It's not obtaining the private key information, only public portion of the key which can be used in the validation process.

**Comment by saleelpk95** [comment]
That answers my query. Thanks a lot!!
And I really hope Azure KeyVault becomes capable of performing the hash on its own.

Closing the issue.

**Comment by saleelpk95** [comment]
@jmhardison 
I was able to get this working for a non-HSM key vault. Thanks for the support!!
I'd like to add something I learned during this task and wanted to check your thoughts on the same.

I couldn't stop wondering what the below line did, and why we had to set the value of the key using the KeyBundle we retrieved from Key Vault. 
`signedXml.SigningKey = key;`
I checked the signing section of the code and the SigningKey was not being used. My guess is that SigningKey is used only when we have to sign the assertion using the default `ComputeSignature()` function. Since we have a custom method for this purpose which relies on KeyVault, I decided to commented out the `// signedXml.SigningKey = key;` line and the code worked.

Again this is not an issue, but something that was bugging my mind.
Appreciate your help!!
Thanks
